README.md
# Open Cloud Robot Table Organization Challenge (OCRTOC)
This is the OCRTOC software package. For more information on OCRTOC please visit the homepage: [www.ocrtoc.org](http://www.ocrtoc.org)

To take part in OCRTOC, you need to develop your own solution using this software package. After uploading your solution to the competition platform, the performance of your solution will be evaluated.

For the simulation stage, we support two simulators: Gazebo and Sapien. You can choose either of them to test your solution on your local machine. On the competition platform we use both simulators to evaluate your solution. As long as your solution works fine with either of the two simulators, it is good enough for the qualification of the real robot stage.

For the real robot stage, hardware drivers will be provided in this software package (around the end of August). Your solution will be tested on real robot hardware. The software interfaces for sensor readings and robot control are the same for both simulation and the real robot hardware. So you will not encounter interface issues when transferring your solution from simulation to the real robot.


## The docker image
In order for your solution to be executable both on your local machine and our competition platform (with cloud computing), we provide you a docker image with a lot of pre-installed software. You need to develop and test your solution within this docker image.

**Operation system**

We highly recommend you to use Ubuntu 18.04 as your operation system, because the provided software package and docker image are guaranteed to work on this OS.

**Install dependencies**

1. Install docker engine on Ubuntu. [https://docs.docker.com/engine/install/ubuntu/](https://docs.docker.com/engine/install/ubuntu/)
1. Install Nvidia container runtime [https://www.celantur.com/blog/run-cuda-in-docker-on-linux/](https://www.celantur.com/blog/run-cuda-in-docker-on-linux/)

**Get the docker image**

The docker image is not part of the OCRTOC software package and needs be downloaded separately. You can use either of the following addresses to download the docker image:

```bash
# Mirror address of China (ShangHai)
sudo docker pull registry.cn-shanghai.aliyuncs.com/tcc-public/ocrtoc:release1.3
# Mirror address of the United States (Silicon Valley)
sudo docker pull registry.us-west-1.aliyuncs.com/tcc-public/ocrtoc:release1.3
```

The content of the docker image is as follows:

- Operation System: Ubuntu 18.04
- ROS melodic-desktop-full
- CUDA 10.2
- Gazebo 9.14.0
- Sapien 0.6
- A lot of simulated scenes for testing purposes, available at /root/ocrtoc_materials/scenes
- Object mesh models with texture information, available at /root/ocrtoc_materials/models



## File Structure of the OCRTOC software package

- **description**: The description files for hardware simulation.
- **docker**: The Dockerfile and build script. You can modify the Dockerfile and rebuild your docker image.
- **gazebo_simulator**: Setups and scripts for the Gazebo simluator.
- **sapien_simluator**: Setups and scripts for the Sapien simluator.
- **ocrtoc_task**: Scripts for task execution and evaluation. (**No modifiction allowed**)
- **ocrtoc_solution**: Setups and scripts for building the solution. Sample code is given in ocrtoc_solution/scripts/commit_solution.py (**Do not change the file name!**). You can modify this file to develop your own solution. In addition you can also add any new software modules to this folder. Make sure that your commit_solution.py serves as the main function of your solution and it incorporates all the needed software modules to do the task. After your solution is uploaded to the competition platform, we will run your solution by ocrtoc_solution/launch/commit_solution.launch (**Do not change the file name!**).

## Interfaces
**ROS topic**

- /clock
- /tf*
- /arm_controller/*
- /gripper_controller/*
- /kinect/*
- /realsense/*
- /joint_group_position_controller/command
- /joint_states

**ROS service**

- /arm_controller/query_state
- /controller_manager/*

**Note**: In the simulation contest of OCRTOC, you are not allowed to use simulator topics to read the 6D pose of the objects in the simulated scenes. You must perform perception using the given object models and visual sensor readings. **Violation of this rule leads to disqualification**.


## Task definition

The target configuration of a task is defined as a list of 6D poses in a given coordinate system (frame) with corresponding object models and scale factors.

In general, we manage tasks using the ROS actionlib functionality. The script ocrtoc_solution/scripts/commit_solution.py will receive tasks in form of a callback function. Each time when a task is received, the corresponding callback function (execute_callback in commit_solution.py) is called.


**Task format**

```
string scene_id                   // the id of the loaded scene
string frame_id                   // the coordinate system, in which the 6D poses of the objects are defined
string[] object_list              // the list of object model names, by which you can get the model from the file system
float64[] scale_list              // the list of scale factors, using which the object model is scaled isotropically to generate the scene, typical values are: 0.5, 1, 2
geometry_msgs/Pose[] pose_list    // the list of 6D object poses (in x,y,z and quaternion) defined in frame_id
  geometry_msgs/Point position
    float64 x
    float64 y
    float64 z
  geometry_msgs/Quaternion orientation
    float64 x
    float64 y
    float64 z
    float64 w
---
# result definition
string status
---
# feedback
string text
```
**Your solution**

After you obtained the task information, you need to implement the core functionality of your solution in the commit_solution.py script to finish the task (object organization). An intuitive way of doing this would be like: 1) recognize the objects in the scene so that you know where to place them; 2) plan the needed motion; 3) execute the planned motion. For some tasks, you may need some kind of active percepion method due to clutters and occlusions in the scene.

**Evaluation**

After you finished the task, you need to publish the actionlib result topic. The format of this result is a string. We do not parse the content of this string. Instead, it is only used to activate our callback function for evaluation. So you can write anything reasonable into this string, such as "done", "finished" and so on. If you do not publish the actionlib result at all, your solution will be terminated after a predefined timeout (e.g. 10 minutes), and then the evaluation will start automatically. We highly recommend you to publish the actionlib result topic, once your solution has finished the execution. This helps us compute the execution time of your solution. For each object in the target configuration, a distance error will be calculated based on its actual pose and its target pose. There will be an upper bound for the distance error of a single object. For each task, distance errors of all the objects are summed up. The performance of the solution is ranked according to the summed distance error of all tasks. If two teams have the same performance, the team consuming less execution time will be ranked higher.



## Use the OCRTOC software package on your local machine

1. **Download the OCRTOC software package**
```bash
# For example, save at $HOME dir.
cd $HOME
git clone git@github.com:OCRTOC/OCRTOC_software_package.git
```

2. **Create a docker container**
```bash
sudo docker run -i -d --gpus all --name ocrtoc_container \
        -e DISPLAY=$DISPLAY -e QT_X11_NO_MITSHM=1 \
        -v /tmp/.X11-unix:/tmp/.X11-unix \
        -v $HOME/OCRTOC_software_package:/root/ocrtoc_ws/src \
        registry.cn-shanghai.aliyuncs.com/tcc-public/ocrtoc:release1.3
sudo xhost +local:`sudo docker inspect --format='{{ .Config.Hostname }}' ocrtoc_container`
```

3. **Launch a simulator and a testing scene**
```bash
# Load a testing scene, e.g. scene 1-1/input.world
sudo docker exec -it ocrtoc_container bash
cd /root/ocrtoc_ws
rm -rf build devel
catkin_make
# You can choose gazebo or sapien for testing
# Start gazebo
roslaunch ocrtoc_task bringup_simulator.launch simulator:=gazebo gui:=true scene:=1-1
# Or start sapien
roslaunch ocrtoc_task bringup_simulator.launch simulator:=sapien gui:=true scene:=1-1
```

4. **Launch your solution**
```bash
sudo docker exec -it ocrtoc_container bash
roslaunch ocrtoc_solution commit_solution.launch
```

5. **Trigger task and evaluation**
```bash
sudo docker exec -it ocrtoc_container bash
# For gazebo
roslaunch ocrtoc_task trigger_and_evaluation.launch simulator:=gazebo scene:=1-1
# For sapien
roslaunch ocrtoc_task trigger_and_evaluation.launch simulator:=sapien scene:=1-1
```

## Submit your solution for the simulation contest
You are required to submit your solution by uploading a docker image to the [competition platform](https://tianchi.aliyun.com/competition/entrance/531815/introduction).

**Please make sure**
- Install all the dependencies into the docker image you submit.
- Disable GUI in your solution, and delete unnecessary print.
- Remove your source code and leave the "catkin_make install" results in you docker image.
- Your solution can be fully launched by the ocrtoc_solution/launch/commit_solution.launch


**Submission example**
```
# 1. Install anything you need in your docker image, such as pytorch.
sudo docker exec -it ocrtoc_container bash
pip install torch torchvision

# 2. After your code is compiled, remove unnecessary files in the container.
cd /root/ocrtoc_ws
catkin_make install
rm -rf build devel

# 3. Verify whether your solution can be launched in the following way
# In terminal 1
sudo docker exec -it ocrtoc_container bash
source /root/ocrtoc_ws/install/setup.bash
roslaunch ocrtoc_task bringup_simulator.launch simulator:=gazebo gui:=true scene:=1-1
# In terminal 2
sudo docker exec -it ocrtoc_container bash
source /root/ocrtoc_ws/install/setup.bash
roslaunch ocrtoc_solution commit_solution.launch
# In terminal 3
sudo docker exec -it ocrtoc_container bash
source /root/ocrtoc_ws/install/setup.bash
roslaunch ocrtoc_task trigger_and_evaluation.launch simulator:=gazebo scene:=1-1

# 4. Export a docker image by the docker container.
sudo docker commit ocrtoc_container your_submission_docker_image_name

# 5.Submit you docker image.
# We will provide a submission instruction on the competition platform later.
```

## Related documentations
[SAPIEN GUI Documentation](sapien_simulator/README.md)

## Known simulator issues
- The reset time button in the GUI of the gazebo simulator should not be used. It will cause the robot to explode. Resetting can be done by ROS service /gazebo/reset_simulation instead.

- In general, the physics engine used in Sapien works more stable than that in Gazebo (open dynamics engine). If you encounter weird issues while using Gazebo, please use Sapien instead.


-----

## Information about the real robot stage
**Hardware safety configuration for the robot arm**
- Joint position limits (min: -pi, max: pi).
- Joint velocity limits (max: 45 deg/s).
- A safety plane is set at 2cm above the table.


**Differences between the real robot stage and the simulation stage**
- The topic_name/frame_id/encoding/camera_info of cameras may changed, check the [camera_image_info.md](https://github.com/OCRTOC/OCRTOC_software_package/blob/master/docs/camera_image_info.md). The follwing topics are recommended. You can also make changes on the camera launch/prarameters.
  - /kinect/color/image_rect_color
  - /kinect/depth_to_color/image_raw
  - /realsense/color/image_raw
  - /realsense/aligned_depth_to_color/image_raw
- Some contents of the tf tree, check the [tf.png](https://github.com/OCRTOC/OCRTOC_software_package/blob/master/docs/tf.png).
- The "arm_controller" won't be launched as the deafult controller. The default controllers are the pos_traj_controller and the scaled_pos_traj_controller. If you are using moveit, you will not encounter the controller issues.
- The control values of /gripper_controller/gripper_cmd/goal. In the real robot stage, "position = 0.0" means openning the gripper to its maximum, and "position = 0.085" means closing the gripper. In addtion, the "max_effort" must be in range 30.0N - 100.0N.

**Sample data collected from the real robot**
- [tf.bag](https://ocrtoc-public.oss-cn-hangzhou.aliyuncs.com/rosbag/1023/tf.bag): A ROS bag including the whole tf tree.
- [realsense.bag](https://ocrtoc-public.oss-cn-hangzhou.aliyuncs.com/rosbag/1023/realsense.bag) : A ROS bag including all topics of realsense.
- [kinect.bag](https://ocrtoc-public.oss-cn-hangzhou.aliyuncs.com/rosbag/1023/kinect.bag) : A ROS bag including all topics of kinect.
- topic_list of arm/cameras/gripper (/docs)

**Important rules for real robot trials**

Before you begin a real robot trial, make sure you follow the following rules:
- (If applicable) Get the latest release of OCRTOC software package and merge it into your code.
- (If applicable) Update your docker image by rebuilding the Dockerfile or by manual installing [dependencies](https://github.com/OCRTOC/OCRTOC_software_package/commit/89a760d4336698661cdbdf39db01658ff9d75d58#diff-ebacf6f6ae4ee68078bb16454b23247d).
- Make sure you use motion planning before you move the robot arm, otherwise your code may result in hardware damage.
- Make sure the joint velocities generated by your code are within the allowed joint velocity limitis (max: 45 deg/s). Otherwise the hardware safety system will stop the execution of the arm movement.

**Procedure of a real robot trial**

For the first real robot trial you need to finish uploading your docker image at least 12 hours before your trial appointment, so that we have enough time to download your docker image. For later trials, you need to finish the uploading at least 2 hours before your trial appointment.

1. Email us the following information:
- The storage link of your docker image
- If you use a private registry, you need to provide us the username and password, so that we can get access to your registry.


2. Before your trial starts, we will create a docker container and launch all needed drivers.

```
# Create container
sudo docker run -i -d --gpus all --name trial_container --network host \
    --privileged -v /dev:/dev -e DISPLAY=$DISPLAY -e QT_X11_NO_MITSHM=1 \
    -v /tmp/.X11-unix:/tmp/.X11-unix ${your_image_name}
sudo xhost +local:`docker inspect --format='{{ .Config.Hostname }}' trial_container`

# Launch drivers.
sudo docker exec -it trial_container bash
source /root/ocrtoc_ws/install/setup.bash
roslaunch ocrtoc_task all_devices.launch robotip:=${robot_ip}
```

3. During the trial, we will launch your solution and triger the task.

```
# Launch solution
docker exec -it trial_container bash
source /root/ocrtoc_ws/install/setup.bash
roslaunch ocrtoc_solution commit_solution_real_stage.launch

# Trigger
docker exec -it trial_container bash
source /root/ocrtoc_ws/install/setup.bash
roslaunch ocrtoc_task trigger_real_stage.launch scene:=1-1
```

4. During the trial, you can watch the whole procedure using zoom (we will provide the link). We will share the output of the terminal and rviz through screen sharing. You can choose the topics that you want to monitor in rviz and put your rviz launch command in "commit_solution_real_stage.launch". 

5. You are allowed to update your docker image during the trial. 

6. You can save the data (up to 5GB) that you want to keep under "/root/upload". After the trial we will send you the data and the log file of your solution.


## Technical support
If you encounter technical issues regarding this software package, please contact us at info@ocrtoc.org


CHANGELOG.md
## [V1.5] release date: 2020-10-23
**Changes in v1.5**
- Removed the ocrtoc_materials_patch folder.
- Added a launch file for MoveIt (/ur5_e_robotiq_2f_85_moveit_config/launch/all.launch).
- Added kinect/realsense/ur5e calibration files (/ocrtoc_task/config)
- Updated urdf files for the real robot stage (description/ur_e_description/urdf). The "ur5e_joint_limited_robot.urdf.xacro" file will load the calibration files.
- Updated all models for the real robot stage (/ocrtoc_materials_real_stage/models).
- Updated the kinect/realsense launch files, improved the image quality.
- Added a FAQ document about the issues we encountered in the trial period (/docs).
- Added a document about the detail information of camera topics (/docs).
- Added documents about the topic lists (/docs).
- Extended README.md with "Differences between the real robot stage and the simulation stage" and "Sample data collected from the real robot".

## [V1.4] release date: 2020-10-10
**Changes in v1.4**
- Added ROS drivers (/drivers).
    - Kinect DK driver.
    - UR5E driver.
    - Robotiq-2F85 driver.
    - Realsense D435i driver.
- Added models and scenes for the real robot stage (/ocrtoc_materials_real_stage).
- Added moveit config (/ur5_e_robotiq_2f_85_moveit_config).
- Added launch files and shell scripts (/ocrtoc_task/launch).
- Updated Dockerfile. Added install dependencies.
- Extended README.md with "Information about the real robot stage".

## [V1.3] release date: 2020-09-11
**Changes in v1.3**
- Added Dockerfile and build script (/docker). You can rebuild the docker image as you wish using the Dockerfile.
- Upgraded docker image. Follow the instructions in the readme file to download the new docker image. However this is not a MUST, if the previous docker image already fulfiled your development requirement.
    - Replaced cuda-runtime with cuda-devel.
    - Added models and scenes released in v1.1.
    - Fixed the Sapien controller issue (https://github.com/OCRTOC/OCRTOC_software_package/issues/10)
- Fixed the Sapien camera_info timestamp issue (https://github.com/OCRTOC/OCRTOC_software_package/issues/14)
- For those of you who want to do some fixings (such as gripper or urdf models) on your own, we allow the modifiction in the following folders "/description", "/gazebo_simulator", "/sapien_simluator".

## [V1.2] release date: 2020-09-03
**Changes in v1.2**
- Updated the method for task evaluation. For each task we calculate the average distance error of all objects instead of their sum.
- Fixed the Sapien object inertia issue (https://github.com/OCRTOC/OCRTOC_software_package/issues/6)
- Fixed the Sapien GUI issue (https://github.com/OCRTOC/OCRTOC_software_package/issues/8)
- Fixed the Gazebo gripper issue (https://github.com/OCRTOC/OCRTOC_software_package/issues/9)

## [V1.1] release date: 2020-08-28
**Changes in v1.1**
- Added 20 object models and 100 trial scenes (available at ocrtoc_materials_patch). You can use the script ocrtoc_materials_patch/patch.sh to copy the new models and scenes into your docker image. The new object models will be re-used in the simulation contest and in the real robot stage.
- Added the task evaluation module (search "Trigger task and evaluation" in readme). Your solution will be evaluated by the same software module in the simulation contest.
- Changed ambient light strength from 0.3 to 0.7 in all trial scenes.
- Fixed camera_info of the Kinect camera for the Sapien simulator.
- Fixed gripper issues for the Sapien simulator. Now the robotiq 2f-85 gripper model uses an explicitly-built prismatic joint. Before the fix, the gripper model relied on the kinematics constraints of three revolute joints, which sometimes failed to maintain the mimic joint status due to certain external forces.

**Known issues in v1.1**
- The gripper model in the Gazebo simulator is not fixed. If you experience any weird problems while using Gazebo, please use Sapien instead.




sapien_simulator/scripts/sapien_env.py
#!/usr/bin/python3
import numpy as np
import sys
import xml.etree.ElementTree as ET
from transforms3d.euler import euler2quat
import os
import time

# The path below comes from a docker
sys.path.append("/workspace/sapien/build")
import rospkg
import rospy
import pysapien_ros1.core as sapien
import pysapien_ros1.ros1 as sr

RENDER_HZ = 8


def load_sapien_sdf(sdf_file, scene, table_height):
    model_path = os.getenv('SAPIEN_MODEL_PATH')
    assert model_path, 'SAPIEN_MODEL_PATH environment variable is required'
    if model_path[-1] != '/':
        model_path += '/'
    sdf = ET.parse(sdf_file).getroot()
    world = sdf.find('world')
    actors = []
    for l in world.findall('light'):
        assert l.attrib['type'] == 'point'
        color = [float(x) / 3.14 for x in l.find('diffuse').text.split()]
        position = np.array([float(x) for x in l.find('pose').text.split()][:3])
        position[2] += table_height
        scene.add_point_light(position, color)
    for sdf_model in world.findall('model'):
        builder = scene.create_actor_builder()
        sdf_link = sdf_model.find('link')
        sdf_pose = sdf_model.find('pose')
        sdf_inertial = sdf_link.find('inertial')
        assert sdf_inertial is not None
        cs = sdf_link.findall('collision')
        vs = sdf_link.findall('visual')
        for col in cs:
            sdf_geom = col.find('geometry')
            sdf_mesh = sdf_geom.find('mesh')
            sdf_uri = sdf_mesh.find('uri')
            sdf_scale = sdf_mesh.find('scale')
            assert sdf_uri is not None and sdf_scale is not None
            filename = sdf_uri.text.replace('model://', model_path)
            scale = [float(x) for x in sdf_scale.text.strip().split()]
            assert len(scale) == 3
            assert os.path.isfile(filename), filename
            friction = float(col.find('surface').find('friction').find('ode').find('mu').text)
            assert friction == 0.5  # will all be 0.5
            builder.add_multiple_convex_shapes_from_file(filename, scale=scale)
        for v in vs:
            sdf_geom = v.find('geometry')
            sdf_mesh = sdf_geom.find('mesh')
            sdf_uri = sdf_mesh.find('uri')
            sdf_scale = sdf_mesh.find('scale')
            assert sdf_uri is not None and sdf_scale is not None
            filename = sdf_uri.text.replace('model://', model_path)
            scale = [float(x) for x in sdf_scale.text.strip().split()]
            assert len(scale) == 3
            assert os.path.isfile(filename), filename
            builder.add_visual_from_file(filename, scale=scale)
        sdf_mass = sdf_inertial.find('mass')
        sdf_pose = sdf_inertial.find('pose')
        sdf_inertia = sdf_inertial.find('inertia')
        assert sdf_mass is not None and sdf_pose is not None and sdf_inertia is not None
        mass = float(sdf_mass.text)
        xyzrpy = [float(x) for x in sdf_pose.text.strip().split()]
        assert len(xyzrpy) == 6
        ixx = float(sdf_inertia.find('ixx').text)
        iyy = float(sdf_inertia.find('iyy').text)
        izz = float(sdf_inertia.find('izz').text)
        ixy = float(sdf_inertia.find('ixy').text)
        ixz = float(sdf_inertia.find('ixz').text)
        iyz = float(sdf_inertia.find('iyz').text)
        assert ixy == ixz == iyz == 0
        builder.set_mass_and_inertia(mass, sapien.Pose(xyzrpy[:3], euler2quat(*xyzrpy[3:])), [ixx, iyy, izz])
        model_pose = sdf_model.find('pose')
        model = builder.build(name=sdf_model.attrib['name'])
        xyzrpy = np.array([float(x) for x in model_pose.text.strip().split()])
        xyzrpy[2] += table_height
        model.set_pose(sapien.Pose(xyzrpy[:3], euler2quat(*xyzrpy[3:])))
        model.set_velocity([0, 0, 0])
        model.set_damping(1, 1)
        actors.append(model)
    return actors


def setup_table(scene: sapien.Scene, height, table_physical_material):
    table_size = np.array([1, 0.8, 0.01]) / 2
    table_pose = np.array([0, 0, height - 0.01])
    table_vis_material = sapien.PxrMaterial()
    table_vis_material.roughness = 0.025
    table_vis_material.specular = 0.95
    table_vis_material.metallic = 0.6
    rgbd = np.array([171, 171, 171, 255])
    table_vis_material.set_base_color(rgbd / 255)
    builder = scene.create_actor_builder()
    builder.add_box_visual_complex(sapien.Pose(table_pose), table_size, table_vis_material)
    builder.add_box_shape(sapien.Pose(table_pose), table_size, table_physical_material)
    table = builder.build_static("table")
    table.set_pose(sapien.Pose([0, 0, 0], [-0.7071, 0, 0, 0.7071]))

    table_leg_position1 = [0.45, 0.35, height / 2]
    table_leg_position2 = [-0.45, -0.35, height / 2]
    table_leg_position3 = [-0.45, 0.35, height / 2]
    table_leg_position4 = [0.45, -0.35, height / 2]
    table_leg_size = np.array([0.025, 0.025, height / 2 - 0.01])
    builder = scene.create_actor_builder()
    builder.add_box_visual_complex(sapien.Pose(table_leg_position1), table_leg_size)
    builder.add_box_visual_complex(sapien.Pose(table_leg_position2), table_leg_size)
    builder.add_box_visual_complex(sapien.Pose(table_leg_position3), table_leg_size)
    builder.add_box_visual_complex(sapien.Pose(table_leg_position4), table_leg_size)
    legs = builder.build_static("table_leg")
    legs.set_pose(table.get_pose())

    return [table, legs]


def main():
    # Parse ROS path and args
    materials_path = rospy.get_param('~materials_dir',
                                     '/root/ocrtoc_materials')
    args = parse_arg()
    print(args)
    if args.paused and not args.gui:
        raise RuntimeError(
            "Argument paused is only useful when GUI is activated. It is only for debug purpose. "
            "Your program will directly end when using paused:=true with gui:=false")

    current_path = rospkg.RosPack().get_path('sapien_simulator')
    engine = sapien.Engine()
    optifuser_config = sapien.OptifuserConfig()
    optifuser_config.use_shadow = False
    renderer = sapien.OptifuserRenderer(glsl_dir=os.path.join(current_path, "./glsl_shader/130"),
                                        glsl_version="130",
                                        config=optifuser_config)
    engine.set_renderer(renderer)
    controller = sapien.OptifuserController(renderer)

    # Load scene and ground
    scene_config = sapien.SceneConfig()
    scene_config.solver_iterations = 25
    scene_config.solver_velocity_iterations = 2
    scene_config.enable_pcm = False
    scene_config.default_restitution = 0
    scene_config.default_dynamic_friction = 0.5
    scene_config.default_static_friction = 0.5
    scene = engine.create_scene(scene_config)
    scene.set_timestep(1 / 250)
    ground_material = sapien.PxrMaterial()
    ground_color = np.array([202, 164, 114, 256]) / 256
    ground_material.set_base_color(ground_color)
    ground_material.specular = 0.5
    scene.add_ground(-0.8, render_material=ground_material)

    if args.gui:
        controller.set_current_scene(scene)
        controller.set_camera_position(2.5, 0, 3)
        controller.set_camera_rotation(3.14, -0.7)
        controller.show_window()

    # Load table
    table_height = 0.0

    # Load sdf
    os.environ.update({
        "SAPIEN_MODEL_PATH": os.path.join(materials_path, "models")})
    sdf_objects = load_sapien_sdf(args.world_name, scene, table_height)

    # scene.set_shadow_light([0, -1, -1], [1, 1, 1])
    scene.set_ambient_light((0.5, 0.5, 0.5))

    sr.init_spd_logger()
    scene_manager = sr.SceneManager(scene, "")
    loader = scene_manager.create_robot_loader()
    loader.fix_root_link = True
    gripper_material = engine.create_physical_material(1.2, 0.8, 0.01)
    urdf_config = {
        "link": {
            "robotiq_2f_85_left_pad": {"material": gripper_material, "patch_radius": 0.02, "min_patch_radius": 0.005},
            "robotiq_2f_85_right_pad": {"material": gripper_material, "patch_radius": 0.02,
                                        "min_patch_radius": 0.005}}}

    # Load robot
    robot, manager = loader.load_from_parameter_server("", urdf_config, 125)
    init_qpos = np.array([-1.57, -1.57, 1.57, 0, 0, 0, 0, 0, 0, 0, 0, 0])
    robot.set_qpos(init_qpos)
    robot.set_drive_target(init_qpos)
    manager.set_drive_property(3000, 500, 1000, [0, 1, 2, 3, 4, 5])
    manager.set_drive_property(200, 50, 300, [6, 7])
    manager.set_drive_property(100, 40, 300, [8, 9, 10, 11])
    scene_manager.start_all_ros_camera(30)
    scene_manager.start_get_model_service("/sapien/get_model_state", sdf_objects)
    scene.step()

    # Start
    start_time = time.time()
    scene_manager.start()
    step = 0
    timestep = scene.get_timestep()
    next_step_time = time.time() + timestep
    mimic_joints = robot.get_active_joints()[20:24]
    if args.gui:
        if args.paused:
            while not controller.should_quit:
                scene.update_render()
                controller.render()

            while True:
                step_and_render(manager, scene, controller, step, next_step_time)
                mimic_joint(robot, mimic_joints)
                next_step_time += timestep
                step += 1
        else:
            while not controller.should_quit:
                step_and_render(manager, scene, controller, step, next_step_time)
                mimic_joint(robot, mimic_joints)
                next_step_time += timestep
                step += 1
    else:
        try:
            while True:
                step_only(manager, scene, next_step_time)
                mimic_joint(robot, mimic_joints)
                next_step_time += timestep
                step += 1
        except KeyboardInterrupt:
            print("Simulation stopped by user")


scene = None


def step_and_render(manager, scene, controller, step, next_step_time):
    manager.balance_passive_force()
    now = time.time()
    while now < next_step_time:
        time.sleep(1e-4)
        now = time.time()
    scene.step()
    scene.update_render()
    if step % RENDER_HZ == 0:
        controller.render()


def step_only(manager, scene, next_step_time):
    manager.balance_passive_force()
    now = time.time()
    while now < next_step_time:
        time.sleep(1e-4)
        now = time.time()
    scene.step()
    scene.update_render()


def mimic_joint(robot, mimic_joints):
    left_target = robot.get_qpos()[7] * 17.86
    right_target = robot.get_qpos()[6] * 17.86
    mimic_joints[0].set_drive_target(right_target)
    mimic_joints[1].set_drive_target(right_target)
    mimic_joints[2].set_drive_target(left_target)
    mimic_joints[3].set_drive_target(left_target)


def parse_arg():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true", help="show gui for visualization")
    parser.add_argument("--paused", action="store_true", help="start simulator in a paused mode")
    parser.add_argument("--world_name", type=str, help="scene name for loading")
    print(sys.argv[1:-2])
    return parser.parse_args(sys.argv[1:-2])


if __name__ == '__main__':
    import sys

    sr.ros_init("iros_pipeline", sys.argv)

    main()
